port/stack_trace.cc

```diff
@@ -135,6 +137,68 @@ void PrintStack(void* frames[], int num_frames) {
 }
 
 void PrintStack(int first_frames_to_skip) {
+#ifdef ROCKSDB_DLL
+  // LIB_MODE=shared build produces modiocre information from the above
+  // backtrace+addr2line stack trace method. Try to use GDB in that case.
+  bool dll = true;
+#else
+  bool dll = false;
+#endif
+  // Also support invoking interactive debugger on stack trace, with this
+  // envvar set to non-empty
+  char* debug_env = getenv("ROCKSDB_DEBUG");
+  bool debug = debug_env != nullptr && strlen(debug_env) > 0;
+
+  if (dll || debug) {
+    // Allow ouside debugger to attach, even with Yama security restrictions
+#ifdef PR_SET_PTRACER_ANY
+    (void)prctl(PR_SET_PTRACER, PR_SET_PTRACER_ANY, 0, 0, 0);
+#endif
+    // Try to invoke GDB, either for stack trace or debugging
+    char pid_str[20];
+    snprintf(pid_str, sizeof(pid_str), "%lld", (long long)getpid());
+    pid_t child_pid = fork();
+    if (child_pid == 0) {
+      // child process
+      if (debug) {
+        fprintf(stderr, "Invoking GDB for debugging (ROCKSDB_DEBUG=%s)...\n",
+                debug_env);
+        execlp(/*cmd in PATH*/ "gdb", /*arg0*/ "gdb", "-p", pid_str,
+               (char*)nullptr);
+        return;
+      } else {
+        fprintf(stderr, "Invoking GDB for stack trace...\n");
+        // TODO: does this always select the correct thread in gdb?
```

Comment:
From my limited manual testing, it seems that main thread's backtrace is always printed even when assertion failure is in compaction threads. Researching on the internet does not give a solution to attach to the failing thread automatically. I wonder if it still makes sense to change default to shared library for CI tests as the backtrace may give limited information (the old method prints backtrace for the failing thread).